test/Microsoft.Azure.WebJobs.Host.UnitTests/Loggers/FunctionExecutorLogOrdering.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Description;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Host.Loggers;
+using Microsoft.Azure.WebJobs.Host.TestCommon;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Host.UnitTests.Loggers
+{
+    // Test that FunctionExecutor emits logs in the right order
+    public class FunctionExecutorLogOrdering
+    {
+        public const string PreBindLog = "[Start]";
+        public const string ParamLog = "[Param]";
+        public const string ParamFailLog = "[ParamFail]"; // parameter binding failed
+        public const string PostBindLog = "[PostBind]";
+        public const string BodyLog = "Body";
+        public const string BodyFailLog = "[BodyFail]";
+        public const string CompletedLog = "[Done]";
+        public const string CompletedFailLog = "[DoneFail]";
+
+        // Test that instrumentation points are emitted in the right order.
+        [Fact]
+        public void TestSuccess()
+        {
+            var logger = new MyLogger();
+            var host = TestHelpers.NewJobHost<MyProg>(logger, new TestExt());
+            host.Call("test");
+
+            Assert.Equal(PreBindLog + ParamLog + PostBindLog + BodyLog + CompletedLog, logger._log.ToString());
+        }
+
+        [Fact]
+        public void TestWithFailParam()
+        {
+            var logger = new MyLogger();
+            var host = TestHelpers.NewJobHost<MyProg>(logger, new TestExt());
+
+            Assert.Throws<FunctionInvocationException>(() => host.Call("testFailParam"));
+
+            Assert.Equal(PreBindLog + ParamFailLog + PostBindLog + CompletedFailLog, logger._log.ToString());
+        }
+
+        [Fact]
+        public void TestWithFailBody()
+        {
+            var logger = new MyLogger();
+            var host = TestHelpers.NewJobHost<MyProg>(logger, new TestExt());
+            Assert.Throws<FunctionInvocationException>(() => host.Call("testFailBody"));
+
+            Assert.Equal(PreBindLog + ParamLog + PostBindLog + BodyFailLog + CompletedFailLog, logger._log.ToString());
+        }
+
+        public class MyLogger : IAsyncCollector<FunctionInstanceLogEntry>
+        {
+            public StringBuilder _log = new StringBuilder();
+
+            string key = "k1";
+
+            public Task AddAsync(FunctionInstanceLogEntry item, CancellationToken cancellationToken = default(CancellationToken))
+            {
+                if (item.IsStart)
+                {
+                    Assert.Null(item.EndTime);
+                    _log.Append(PreBindLog);
+
+                    Assert.False(item.Properties.ContainsKey(key));
+                    item.Properties[key] = 1;
+                }
+                else
```

Comment:
nit -- you can make these `else if (item.IsPostBind)` / `else if (item.IsComplete)` now, and probably throw at the end if it doesn't match any?